### PR TITLE
[SignalR] Implement IConnectionLifetimeFeature

### LINKED ISSUE #5329: Implement IConnectionLifetimeFeature
This feature allows the application to forcefully abort the underlying transport without going through the pipe. It's implemented in Kestrel and should be implemented in the transport layer it should be implemented at the http.connections layer.

### LINKED ISSUE #19367: Hub: Closing the browser on a long running task does not trigger OnDisconnectedAsync until after the task has finished
### Describe the bug

Closing the browser on a long running task does not trigger OnDisconnectedAsync until after the task has finished.
What I'm after is being able to cancel a normal hub method on user disconnect/browser close.

### To Reproduce
``` C#
 public class TestHub : Hub<IProgressHub>
    {

        public override Task OnDisconnectedAsync(Exception exception)
        {
            // This is triggered only after the task has finished.
            return base.OnDisconnectedAsync(exception);
        }


        public async Task<bool> RunTestAsync()
        {
            return (await _TestTaskAsync(this.Context.ConnectionAborted));
        }

        private async Task<bool> _TestTaskAsync(CancellationToken cancel)
        {
            for (int i = 0; i < 1000; i++)
            {
                if (cancel.IsCancellationRequested)
                {
                    // never triggered.
                }

                await Task.Delay(1000);
                await this.Clients.Client(this.Context.ConnectionId).ReportProgress(i.ToString());
            }

            return true;
        }
    }
```

``` C#
 public interface IProgressHub
    {
        Task ReportProgress(string message);
    }
```

### Further technical details
.NET Core SDK (reflecting any global.json):
 Version:   3.1.101
 Commit:    b377529961

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.101\

Host (useful for support):
  Version: 3.1.1
  Commit:  a1388f194c

.NET Core SDKs installed:
  2.1.802 [C:\Program Files\dotnet\sdk]
  3.1.101 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/5329 - Implement IConnectionLifetimeFeature
Fixes https://github.com/dotnet/aspnetcore/issues/19367 - Trigger Context.ConnectionAborted sooner

### CODE REVIEW

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs [hunk 1]
```diff
@@ -170,6 +176,13 @@ public IDuplexPipe Application
 
         public HttpContext HttpContext { get; set; }
 
+        public override CancellationToken ConnectionClosed { get; set; }
+
+        public override void Abort()
+        {
+            ThreadPool.UnsafeQueueUserWorkItem(cts => ((CancellationTokenSource)cts).Cancel(), _connectionClosedTokenSource);
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:183]: Abort should be using HttpContext.Abort (if it's not the fake HttpContext). So even if you keep this, it should result in killing the transport directly.

Comment (halter73) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:183]: I'm not sure if calling HttpContesxt.Abort() is the way to do it, but I agree this should also stop the transport if it hasn't stopped already.

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:183]: > this should also stop the transport if it hasn't stopped already.

That's the part I don't have currently

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:183]: Is there a race with HttpContext.Abort and Kestrel writing the bytes in the response stream to the network? As soon as I added this it started randomly failing tests that were looking for a close message.

Comment (halter73) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:183]: I wouldn't call it a race as much as behaving as intended. Aborting drops any unflushed bytes in Kestrel's internal buffers. If you want to guarantee delivery of the bytes you've written, you cannot abort. This is the same reason why we cannot reliably abort a SignalR connection and send a CloseMessage.

> I'm not sure if calling HttpContesxt.Abort() is the way to do it

Apparently it isn't.

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:183]: So do we need to abort the HttpContext? Or can we assume the server will clean up correctly when we return from the middleware?

Comment (halter73) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:183]: The server should do the right thing to complete the request and/or close the connection when middleware completes. If the client receives data slower the configured rate limit, the server will abort the connection.

**Comment by BrennanConroy** [comment]
I think this is better now. Before I was calling Abort, but that should be called by the server/someone else instead.

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs [hunk 2]
```diff
@@ -313,6 +331,9 @@ private async Task WaitOnTasks(Task applicationTask, Task transportTask, bool cl
                     Application?.Output.Complete(transportTask.Exception?.InnerException);
                     Application?.Input.Complete();
 
+                    // Trigger ConnectionClosed
+                    _connectionClosedTokenSource.Cancel();
```
Comment (halter73) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:335]: We should dispatch to the thread pool before calling Cancel().

File: src/SignalR/server/Core/src/HubConnectionContext.cs [hunk 3]
```diff
@@ -65,7 +65,7 @@ public HubConnectionContext(ConnectionContext connectionContext, HubConnectionCo
 
             _connectionContext = connectionContext;
             _logger = loggerFactory.CreateLogger<HubConnectionContext>();
-            ConnectionAborted = _connectionAbortedTokenSource.Token;
+            ConnectionAborted = CancellationTokenSource.CreateLinkedTokenSource(_connectionAbortedTokenSource.Token, connectionContext.ConnectionClosed).Token;
```
Comment (halter73) [inline src/SignalR/server/Core/src/HubConnectionContext.cs:68]: Why doesn't the ConnectionClosed token firing abort the HubConnectionContext?

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs (hunk 2 above)
Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs:335]: Ah true, good catch

File: src/SignalR/server/Core/src/HubConnectionContext.cs (hunk 3 above)
Comment (BrennanConroy) [reply src/SignalR/server/Core/src/HubConnectionContext.cs:69]: It does now

Comment (halter73) [reply src/SignalR/server/Core/src/HubConnectionContext.cs:69]: Doesn't that make it redundant to link the ConnectionClosed token here?

Comment (BrennanConroy) [reply src/SignalR/server/Core/src/HubConnectionContext.cs:69]: Ah, because now we call Abort which cancels the token :+1:

Comment (davidfowl) [reply src/SignalR/server/Core/src/HubConnectionContext.cs:69]: This linking of tokens makes me unhappy, do we need it?

Comment (BrennanConroy) [reply src/SignalR/server/Core/src/HubConnectionContext.cs]: Removed

File: src/SignalR/server/Core/src/HubConnectionContext.cs [hunk 4]
```diff
@@ -630,6 +632,8 @@ private static void AbortConnection(object state)
                 {
                     connection._streamTracker.CompleteAll(new OperationCanceledException("The underlying connection was closed."));
                 }
+
+                connection._closedRegistration.Dispose();
```
Comment (halter73) [inline src/SignalR/server/Core/src/HubConnectionContext.cs:636]: Does this always get run when the HubConnection completes? Can we just dispose the HubConnectionContext at then end of HubConnectionHandler.OnConnectedAsync?

Comment (BrennanConroy) [reply src/SignalR/server/Core/src/HubConnectionContext.cs:636]: > Does this always get run when the HubConnection completes?

`HubConnectionContext`* and yes it always runs.

> Can we just dispose the HubConnectionContext at then end of HubConnectionHandler.OnConnectedAsync?

We could (although probably still keep it above `_lifetimeManager.OnDisconnectedAsync`), any strong reason to do that?

Comment (halter73) [reply src/SignalR/server/Core/src/HubConnectionContext.cs:636]: I think it's probably fine then.

When I was looking at this earlier, it appeared that AbortConnection() is only triggered by AbortAllowReconnect() which seems to only be called in non-graceful circumstances. How does this get called when the client gracefully closed via HubConnection.StopAsync()?

Comment (BrennanConroy) [reply src/SignalR/server/Core/src/HubConnectionContext.cs:636]: It's called from `HubOnDisconnectedAsync`
https://github.com/dotnet/aspnetcore/blob/fa20111e47774b6c08f82a1f0c6373cc30e99a83/src/SignalR/server/Core/src/HubConnectionHandler.cs#L195

Although, if the lifetimemanager throws in `OnConnectedAsync` this wouldn't be disposed which would give a good reason to move it.

**Review by halter73 — APPROVED** [review]

**Pull request merged**